pkg/storegateway/bucket_chunk_reader.go

```diff
@@ -126,88 +131,66 @@ func (r *bucketChunkReader) loadChunks(ctx context.Context, res []seriesChunks,
-		// There is also crc32 after the chunk, but we ignore that.
-		chunkLen = n + 1 + int(chunkDataLen)
-		if chunkLen <= len(cb) {
-			err = populateChunk(&(res[pIdx.seriesEntry].chks[pIdx.chunkEntry]), rawChunk(cb[n:chunkLen]), chunksPool)
-			if err != nil {
-				return errors.Wrap(err, "populate chunk")
-			}
-			localStats.chunksTouched++
-			localStats.chunksTouchedSizeSum += chunkLen + crc32.Size
-			continue
-		}
-
-		// Read entire chunk into new buffer.
-		// TODO: readChunkRange call could be avoided for any chunk but last in this particular part.
-		nb, err := r.block.readChunkRange(ctx, seq, int64(pIdx.offset), int64(chunkLen), []byteRange{{offset: 0, length: chunkLen}})
+		chunkDataLen, err := binary.ReadUvarint(reader)
 		if err != nil {
-			return errors.Wrapf(err, "preloaded chunk too small, expecting %d, and failed to fetch full chunk", chunkLen)
+			return errors.Wrap(err, "parsing reader")
 		}
-		if len(*nb) != chunkLen {
-			return errors.Errorf("preloaded chunk too small, expecting %d", chunkLen)
+		// We ignore the crc32 after the chunk data.
+		chunkEncDataLen := int(chunkDataLen) + chunks.ChunkEncodingSize
+		cb := chunksPool.Get(chunkEncDataLen)
+
+		fullyRead, err := io.ReadFull(reader, cb)
+		// We get io.EOF when there are 0 bytes read and io.UnexpectedEOF when there are some but not all read.
+		if errors.Is(err, io.EOF) || errors.Is(err, io.ErrUnexpectedEOF) {
+			if chunksLeft := len(pIdxs) - 1 - i; chunksLeft != 0 {
+				// Unexpected EOF for last chunk could be a valid case if we have underestimated the length of the chunk.
+				// Any other errors are definitely unexpected.
+				return fmt.Errorf("underread with %d more remaining chunks in seq %d start %d end %d", chunksLeft, seq, part.Start, part.End)
+			}
+			if err = r.fetchChunkRemainder(ctx, seq, int64(reader.offset), int64(chunkEncDataLen-fullyRead), cb[fullyRead:], &localStats); err != nil {
+				return errors.Wrapf(err, "refetching chunk seq %d offset %x length %d", seq, pIdx.offset, pIdx.length)
+			}
+		} else if err != nil {
+			return errors.Wrap(err, "read chunk")
 		}
 
-		localStats.chunksRefetched++
-		localStats.chunksRefetchedSizeSum += len(*nb)
-		err = populateChunk(&(res[pIdx.seriesEntry].chks[pIdx.chunkEntry]), rawChunk((*nb)[n:]), chunksPool)
+		err = populateChunk(&(res[pIdx.seriesEntry].chks[pIdx.chunkEntry]), rawChunk(cb))
 		if err != nil {
-			r.block.chunkPool.Put(nb)
 			return errors.Wrap(err, "populate chunk")
 		}
 		localStats.chunksTouched++
-		localStats.chunksTouchedSizeSum += chunkLen + crc32.Size
+		localStats.chunksTouchedSizeSum += chunkEncDataLen
+	}
+	return nil
+}
 
-		r.block.chunkPool.Put(nb)
+func (r *bucketChunkReader) fetchChunkRemainder(ctx context.Context, seq int, offset, length int64, dest []byte, localStats *queryStats) error {
+	refetchReader, err := r.block.chunkRangeReader(ctx, seq, offset, length)
```

Comment:
Here we expect that the `dest` length is equal to the input `length`. If it's not, we'll have a silent bug because `io.ReadFull()` ignores `length` and just fill `dest`. For this reason, I would add a check at the beginning of the function to ensure consistency between `len(dest)` and `length`.